### ISSUE TITLE
Add Block: Search

### ISSUE THREAD

**Issue by melchoyce** [WordPress/gutenberg#1800]
## Attributes

- Display search icon

## States

Selected:

![search](https://user-images.githubusercontent.com/2846578/27979861-2128ff1e-6347-11e7-96ae-33b567d955fb.png)

Neutral:

![search block](https://user-images.githubusercontent.com/2846578/27979866-253d9a24-6347-11e7-825b-648b4d9e24f8.png)

-----

Question:
- I added "Display search icon," but should that be a theme decision, not a user decision?

**Comment by StaggerLeee** [comment]
I hope all those native "widgets" in editor will be hidden on default. Or easy hidden with code.

With them I cannot say anymore to beginners Users, just go ahead in backend, you cannot ruin layout, design, or website, zero chance. 
I set backend in an way they cannot ruin anything, even if they tried it by purpose.

**Comment by paulwilde** [comment]
> I added "Display search icon," but should that be a theme decision, not a user decision?

I believe it should be up to the theme in this case, and the option could be added (hopefully) by extending this particular block.

**Comment by afercia** [comment]
Ideally, placeholders should not be used as replacement for labels, see https://core.trac.wordpress.org/ticket/40331 especially in complex forms with several fields.

We've discussed this issue a few times in the accessibility team and agreed it could be acceptable to use a placeholder as in the "Neutral" mockup above for single form fields with a very specific task to accomplish. Even in this case, the search field will need a visually hidden label because the placeholder is going to [not contribute any longer to the computation of the accessible name](https://core.trac.wordpress.org/ticket/40331#comment:3).

**Comment by melchoyce** [comment]
Thanks for the feedback — I'll work on an update.

**Comment by karmatosed** [comment]
I would say the search icon should be down to the theme, I am not sure about it as a user decision. My vote would be to let themes decide that.

**Comment by melchoyce** [comment]
Updated to remove the search icon option, and add an editable label and placeholder:

![search v2](https://user-images.githubusercontent.com/2846578/28031722-3a2eff64-6576-11e7-9bde-e937430e140b.png)
![search block v2](https://user-images.githubusercontent.com/2846578/28031724-3ba54a38-6576-11e7-9133-691da67d4457.png)

Thinking the label should be "Search" by default, but still editable. The placeholder can default to empty, unless someone clicks in to add one. Does this work better?

**Comment by karmatosed** [comment]
As the project moves towards merge proposal, we are moving issues that aren't needed for that to projects. This doesn't mean they won't get done, they totally can and that's why we're moving to projects. This allows us to focus on the issues that have to happen to Gutenberg. I am closing this but it will live in projects.

**Comment by melchoyce** [comment]
Updated mockup:

![search_v3_1024](https://user-images.githubusercontent.com/2846578/34996164-c8c6d222-faa6-11e7-8677-5be3e25e9632.png)

**Comment by danielbachhuber** [comment]
Copying @westonruter's comment from https://github.com/WordPress/gutenberg/pull/4501#pullrequestreview-97716621:

> It's a difficult question on whether to try to re-use `get_search_form()`. I think the most important thing is that the preview be faithful to the frontend. This is currently a challenge in Gutenberg for widgets because they have separate implementations between editing (JS) and viewing (PHP). A theme can override the PHP rendering to be something completely different already, which would break the WYSIWYG block editing preview. But this would be less likely than the theme defining a searchform.php. Since the block here is not going to be used in normal theme template contexts but rather in new Gutenberg contexts, I'm inclined to think that the search block you're defining here should break away from trying to re-use searchform.php to instead use a new Gutenberg block template, one that leverages all of the new customizability you are introducing. After all, the Latest Posts widget block is not re-using the Recent Posts widget template.
> 
> In the Gutenberg world, theme templates are going to need to be rethought anyway in favor of a block templating system, I think. (Especially to eliminate duplicating templates across PHP and JS.) So I don't think we should try too hard here to re-use existing WP theme templates.

**Comment by gibrown** [comment]
The top search on wp.com is for "", and the second top is for "search". This occurs because users often click the search button rather than type something into the box first. 

There is an unmerged patch for this in https://core.trac.wordpress.org/ticket/34886 that we should fix as a part of this.
